BitVM Changes

storopoli opened the pull request:
Supersedes #2613.

- Exposes the internal `u8` from `Opcode`
- Defines a `FromStr` implementation for `Opcode`.

tcharding commented:
Any reason you didn't just add the `parse_asm` function to the already existent `ScriptBufExt` trait?

tcharding commented on bitcoin/src/blockdata/script/asm.rs:
```diff
@@ -0,0 +1,160 @@
+#[non_exhaustive]
+pub enum AsmParseErrorKind {
+    /// ASM ended unexpectedly.
+    UnexpectedEOF,
```
```suggestion
    UnexpectedEof,
```

storopoli replied:
Thanks! Done in 9f16317109e7a2530d74267140afc0984ff330d0

tcharding reviewed (commented):
Is the aim of this branch to get merged or just for folk to use as is? If the aim is to get all these changes merged then you want to break the PR up into discreet changes. When I did the original branch I just grabbed all Steven's stuff and threw it up so we had something to look at. Also you'll need to think about everything because I did not, like I said I just copied it over.

storopoli commented:
> Is the aim of this branch to get merged or just for folk to use as is?

Get the PR merged. Will reset and break into discrete commit changes.

> Also you'll need to think about everything because I did not, like I said I just copied it over.

Yes, you're totally right. Will take this as a draft and burn some mind CPU cycles on it.

> Any reason you didn't just add the `parse_asm` function to the already existent `ScriptBufExt` trait?

Oh this makes so much sense. I should have thought of this before. Great suggestion, thanks!

storopoli commented on bitcoin/src/blockdata/script/asm.rs:
```diff
@@ -0,0 +1,160 @@
+/// The different kinds of [`ParseAsmError`]s that can occur.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum AsmParseErrorKind {
```
Yes! Thanks! Done in 9f16317109e7a2530d74267140afc0984ff330d0

tcharding commented:
> I should have thought of this before. 

No stress man, its only because I stare at this code all day.

tcharding commented:
changes to `Opcodes`  - good job you picked this up before we 1.0'ed `units`. I would not have noticed it. Thanks man.

tcharding commented:
Will review later, I need to do it without rushing.

storopoli commented:
You know we can drop all the `from_asm` things and just go with the single commit 1407b0cd33b738657a3bf62e407133772534eebb.

I did a `git grep` and searched in the GitHub UI and `BitVM/BitVM` do not use `ScriptBuf::from_asm` if that simplifies the PR to get it merged soon let me know and I'll drop these changes and go only with the changes in the Opcodes.

storopoli commented:
> You know we can drop all the `from_asm` things and just go with the single commit [1407b0c](https://github.com/rust-bitcoin/rust-bitcoin/commit/1407b0cd33b738657a3bf62e407133772534eebb).
> 
> I did a `git grep` and searched in the GitHub UI and `BitVM/BitVM` doesn't use `ScriptBuf::from_asm` if that simplifies the PR to get it merged soon let me know and I'll drop these changes and go only with the changes in the Opcodes.

Confirmed with @lucidLuckylee neither [`BitVM/BitVM`](https://github.com/BitVM/BitVM) nor [`BitVM/rust-bitcoin-script`](https://github.com/BitVM/rust-bitcoin-script) uses the `ScriptBuf::parse_asm` functionality.

Hence this is a single commit PR that only do the necessary changes to Opcodes.
This will make much simpler to review and merge.

tcharding commented:
Need to update the PR description mate.

OrestisAlpos commented:
Hey folks, thanks for this PR and all the work.

> Confirmed with @lucidLuckylee neither [BitVM/BitVM](https://github.com/BitVM/BitVM) nor [BitVM/rust-bitcoin-script](https://github.com/BitVM/rust-bitcoin-script) uses the ScriptBuf::parse_asm functionality.

The [BitVM/rust-bitcoin-scriptexec](https://github.com/BitVM/rust-bitcoin-scriptexec/blob/5c88966fa6ed718cd5a8a8b4a0722dd1f4fad5fb/src/wasm.rs#L14) repo is using parse_asm() though! Can we bring it back?

OrestisAlpos commented:
Hey folks. I already needed this PR, so I have adapted the following repos to use this version of rust-bitcoin.
- [BitVM/rust-bitcoin-scriptexec](https://github.com/BitVM/rust-bitcoin-scriptexec)
- [BitVM/rust-bitcoin-script](https://github.com/BitVM/rust-bitcoin-script)
- [BitVM/rust-bitcoin-script-stack](https://github.com/BitVM/rust-bitcoin-script-stack)
- [BitVM/rust-esplora-client](https://github.com/BitVM/rust-esplora-client)

After this PR gets merged, I can open the corresponding PRs on these repos. Let me know!

storopoli commented:
@OrestisAlpos I've added back `ScriptBuf::parse_asm`.

Can you help review this PR? Particularly the `ScriptBuf::parse_asm` is not the same as the original `bitvm` branch that `BitVM` repos are using now. I had do change minor things so that it play nice with current `master`. I am also not 100% sure on the 32-bit architecture "pointer-size-gating" in `ScriptBuf::parse_asm`.

But all original `bitvm` branch tests passes.

> Ah, and maybe let's try to stop making breaking changes wherever possible, so the dependencies can stabilize?

Check https://github.com/rust-bitcoin/rust-bitcoin/issues/3166 for more context.

OrestisAlpos commented:
Hey @storopoli , over the next week I will be using `BitVM/BitVM`, so I will let you know if more changes are needed in this PR. For now I came  across this: other dependencies of  `BitVM/BitVM` use secp256k1 0.28, but this PR uses 0.29. This seems to create some incompatibilities. If there is no other way to resolve, we could use v 0.28 here as well.

Glad to help review `ScriptBuf::parse_asm`, but I can probably do that in a week or so.

Question, have you already integrated these changes downstream on `BitVM/BitVM` as well?

storopoli commented:
> Hey @storopoli , over the next week I will be using `BitVM/BitVM`, so I will let you know if more changes are needed in this PR. For now I came across this: other dependencies of `BitVM/BitVM` use secp256k1 0.28, but this PR uses 0.29. This seems to create some incompatibilities. If there is no other way to resolve, we could use v 0.28 here as well.
> 
> Glad to help review `ScriptBuf::parse_asm`, but I can probably do that in a week or so.
> 
> Question, have you already integrated these changes downstream on `BitVM/BitVM` as well?

It is always good to pair `rust-bitcoin` and `secp256k1` versions.
So if `rust-bitcoin` uses whatever `secp256k1` version you should update to that version.

Additionally, `rust-bitcoin` re-exports `secp256k`: 
https://github.com/rust-bitcoin/rust-bitcoin/blob/fcbe970ddb936ae16197426c7fdcf5774f17a523/bitcoin/src/lib.rs#L87-L89

And also you can see at https://docs.rs/bitcoin/latest/bitcoin/#reexports

> Question, have you already integrated these changes downstream on BitVM/BitVM as well?

No, that's why I've asked for help to test them. To be honest I don't know how to do the `patch.crates-io` thing since I don't have write access to `rust-bitcoin`, hence this PR does not have a branch in `rust-bitcoin/rust-bitcoin`, but actually in `storopoli/rust-bitcoin` (a fork).
But I'm planning to make PRs to `BitVM` as soon as this PR hits `master` and then another afterwards when this PR makes into a tagged release.

apoelstra commented on primitives/src/locktime/relative.rs:
```diff
@@ -105,6 +105,23 @@ impl LockTime {
 
+    /// Try to interpret the given number as a relative lock time.
+    #[inline]
+    pub fn from_num(num: i64) -> Option<LockTime> {
```
In fd78f83e7e4f2fb6dcbd5ce8ca555b436973f72b:

How does this differ from `from_consensus` except that it does a conversion to `u32` for you?

storopoli replied:
Removed. Good point!

apoelstra commented on primitives/src/opcodes.rs:
```diff
@@ -71,6 +72,40 @@ macro_rules! all_opcodes {
+
+        impl str::FromStr for Opcode {
+            type Err = ();
+            fn from_str(s: &str) -> Result<Self, Self::Err> {
```
In fd78f83e7e4f2fb6dcbd5ce8ca555b436973f72b:

I really don't like this fromstr impl ... it parses many ambiguous things. I'm not sure what we want to do here but I'm inclined to say "no parsing of opcodes from strings in rust-bitcoin" and make downstream users implement their own.

tcharding replied:
@storopoli FYI we investigated this already recently in #3617.

storopoli replied:
I think this is valid being defined here since it is impossible for a crate outside of `rust-bitcoin` to implement it (Rust's orphan rule).
And please note that `FromStr` returns a `Result` which the user must deal with it.
It is not that big of a footgun, as portrayed.

storopoli replied:
I did add some tests that @tcharding had in #3617 with a typo fix that the second roundtrip (`Debug`) is `s2` and not `s1` again.

apoelstra commented on bitcoin/src/blockdata/script/mod.rs:
```diff
@@ -142,18 +145,60 @@ pub fn write_scriptint(out: &mut [u8; 8], n: i64) -> usize {
+/// [`push_bytes::PushBytes::read_scriptint`] or [`read_scriptint_non_minimal`] instead.
+///
+/// Panics if max_size exceeds 8.
+pub fn read_scriptint_size(v: &[u8], max_size: usize, minimal: bool) -> Result<i64, Error> {
```
In fd78f83e7e4f2fb6dcbd5ce8ca555b436973f72b:

Do we really want to expose this in the public API? It seems like it's only used with two values -- 4 and 5 -- and these ought to be special-cased. The 5-byte version is encountered in two places: (a) when parsing locktimes and (b) whe parsing the output of something like OP_ADD with overflow.

storopoli replied:
Removed.

apoelstra commented on bitcoin/src/blockdata/script/mod.rs:
```diff
@@ -142,18 +145,60 @@ pub fn write_scriptint(out: &mut [u8; 8], n: i64) -> usize {
 }
 
+/// Returns minimally encoded scriptint as a byte vector.
+pub fn scriptint_vec(n: i64) -> Vec<u8> {
```
In fd78f83e7e4f2fb6dcbd5ce8ca555b436973f72b:

This also feels like the wrong API. I'd almost rather just expose `write_scriptint`.

storopoli replied:
Removed

apoelstra commented:
I'm also super hesitant to include `parse_asm` in this library because it canonicalizes an ambiguous string-serialization format which (a) accepts way more things than this library outputs, (b) doesn't obviously accept the stuff that comes out of Bitcoin Core's RPC.

It also doesn't need to be here. The `Opcode` and `Script` types are public. It's easy to implement this outside the crate.

storopoli commented:
Fair arguments.

1. Removed all `from_asm` functionality since we can move this to [`BitVM/rust-bitcoin-scriptexec`](https://github.com/BitVM/rust-bitcoin-scriptexec/blob/5c88966fa6ed718cd5a8a8b4a0722dd1f4fad5fb/src/wasm.rs#L14) (Cc @Robinlinus, @lucidLuckylee).
2. Removed all other parsing functions that aren't methods of `Opcode`/`ScriptBuf` since we can also implement those as helper funs in a `BitVM`-org crate.
3. Kept the public exposure of the internal `u8` in `Opcode`.
4. Kept the `FromStr` implementation for `Opcode`. I think this is valid being defined here since it is impossible for a crate outside of `rust-bitcoin` to implement it (Rust's orphan rule).

tcharding commented:
> primitives: expose opcode's internal byte

Sorry man this is way too brief a message to even consider making a change to the public API of one of the soon-to-be-stabalized crates (`primitives`, `hashes`, `io`, `units`). Making a struct field public is a non-trivial change because it means we have to commit to it.

If you think this change should be made it needs (IMO) to go in its own PR and be fully described and justified.

(FTR, off the top of my head, I do not think it should be public. If you really need the `u8` adding a getter and setter is another way to get at it that is more stabalization friendly.)

Sorry to be a nuisance, I did warn you that this BitVM stuff was not as simple as pushing up the the changes.

tcharding commented on primitives/src/opcodes.rs:
```diff
@@ -71,6 +72,40 @@ macro_rules! all_opcodes {
+                    "OP_TRUE" | "TRUE" => Ok(OP_TRUE),
+                    "OP_FALSE" | "FALSE" => Ok(OP_FALSE),
+                    "OP_NOP2" | "NOP2" => Ok(OP_NOP2),
+                    "OP_NOP3" | "NOP3" => Ok(OP_NOP3),
```
This PR has code in in that we discussed in #3617 and decided not to use however the patch makes no attempt to justify _why_ this is now a valid thing to do. If you think that we were wrong in #3617 please add a PR that fully justifies _why_ this is a good change and makes a case for why the conclusion in #3617 was wrong.

tcharding replied:
And FTR the orphan rule is not a valid reason because if a user wants an ad-hoc string parsing function they can write one (just not implement `FromStr for Opcode`) - we don't want to add a `FromStr` impl that suits one set of users and not another.

storopoli commented:
I am closing this. As you have seen I was able to migrate all the BitVM repos to `rust-bitcoin@0.32.5` by implementing what the `bitvm` branch and this PR in the `BitVM` crates.

Thanks for the feedback and reviews guys!

Cheers :)

tcharding commented:
Nice, that's a win.
